scripts/vm/hypervisor/kvm/nsrkvmbackup.sh

```diff
@@ -0,0 +1,270 @@
+                        log "Success"
+        else
+                        log "Failure"
+                        log -e "\n\tNetworker Service NOT FOUND. Make sure that Networker client is properly installed"
+                        exit 1
+        fi
+
+        log -ne "\t[2] Checking if Networker is running\t"
+        if [[ $(systemctl is-active networker) = *active* ]]; then
+                        log "Success"
+        else
+                        log "Failure"
+                        log -e "\n\tNetworker Service is not running. Investigate Networker logs, startup server and try again"
+                        exit 2
+        fi
+        log -ne "\t[3] Checking Networker DNS Resolution\t"
+        if [[ $(getent hosts "$networkerServer") = *$networkerServer* ]]; then
+                        log "Success"
+        else
+                        log "Failure"
+                        log -e "\n\tNetworker Server cannot be resolved. Backups will most probably fail. Consider adding the ip/hostname to /etc/host or fix DNS resolution"
+                        exit 3
+        fi
+
+        log -ne "\t[4] Checking QEMU / Libvirt Versions \t"
+        hvVersion=$(virsh version | grep hypervisor | awk '{print $(NF)}')
+        libvVersion=$(virsh version | grep libvirt | awk '{print $(NF)}' | tail -n 1)
+        apiVersion=$(virsh version | grep API | awk '{print $(NF)}')
+        if vercomp "$hvVersion" \> 2.1.2; then
+                log -n "Success"
+                log -ne "\t\t [ Libvirt: $libvVersion apiVersion: $apiVersion ]"
+                echo
+
+        else
+                log "Failure"
+                log -e "\n\tYour QEMU version $hvVersion is unsupported. Consider upgrading at least to latest QEMU at branch 2"
+                exit 4
+        fi
+
+
+
+        log "Environment Sanity Checks successfully passed"
+}
+
+
+
+echo "
+Cloudstack B&R Framework - EMC Networker backup script
+Version $version
+"
+
+
+backup_domain() {
+
+        name=$1
+        snapName=$2
+        log  "Preparing snapshots and gathering information for backing up domain $name under snapshot name $snapName"
+        log  "Retention time is $retentionTime"
+
+        declare -A TRGSRC
+        while IFS=',' read -r TARGET SOURCE
+        do
+                if [[ $SOURCE != "-" ]]; then
+                        TRGSRC+=(["$TARGET"]="$SOURCE")
+                fi
+        done < <(virsh domblklist "$name" --details | grep file | grep -v 'cdrom' | grep -v 'floppy' | sed 's/  */,/g' |  cut -d',' -f 4-)
+        diskspec=""
+        for target in "${!TRGSRC[@]}"; do
+          log -e "\tDisk for $target is at ${TRGSRC[${target}]}"
+          diskspec="$diskspec --diskspec $target,snapshot=external"
+          disks="$disks ${TRGSRC[${target}]} "
+        done
+
+        cmd="$(virsh snapshot-create-as --domain "$name" --name "$snapName" --no-metadata --atomic --quiesce --disk-only "$diskspec")"
```

Comment:
general remark - could we do this via java using libvirt-java than using virsh?

Reply:
Since it is running on the kvm agent side i am not sure we can do.